server/src/main/java/org/tctalent/server/service/db/impl/SavedListServiceImpl.java

```diff
@@ -747,6 +752,14 @@ public void associateTaskWithList(User user, TaskImpl task, SavedList list) {
 
         //Now assign tasks to candidates in list (if they do not already have the task actively assigned)
         Set<Candidate> candidates = list.getCandidates();
+        if (task.getName().equals("duolingoTest")) {
+            List<DuolingoCoupon> availableCoupons = couponService.getAvailableCoupons();
+            if (candidates.size() > availableCoupons.size()) {
+                throw new NoSuchObjectException(
+                    availableCoupons.size() + " coupons available, but you need "
+                        + candidates.size() + " coupons for your candidates list.");
+            }
+        }
```

Comment:
I don't think that this check belongs here.

Putting here builds a dependency between the SavedListService and the Coupon service. It also makes this method throw an Exceptions which it never did before (and is undocumented in the interface).

This code which is just an error check really would be much better placed in the caller: TaskAssignmentAdminApi.

That will add a dependency on CouponService to the TaskAssignmentAdminApi - but I think that is preferable to adding it to the already complex and core SavedListService.